Tests/Test.cs

```diff
@@ -0,0 +1,327 @@
+                    break;
+                case (ProductAdded(var (productId, quantity, price), var addedAt), not null):
+                    var toUpdate = current.ProductItems.SingleOrDefault(p => p.ProductId == productId);
+
+                    if (toUpdate == null)
+                    {
+                        current.ProductItems.Add(new ShoppingCartModel.PricedProductItem
+                        {
+                            ProductId = productId, Quantity = quantity, Price = price, AddedAt = addedAt,
+                        });
+                    }
+                    else
+                    {
+                        toUpdate.Quantity += quantity;
+                        toUpdate.UpdatedAt = addedAt;
+                    }
+
+                    break;
+                case (ProductRemoved(var (productId, quantity, price), var removedAt), not null):
+                    var existing = current.ProductItems.Single(p => p.ProductId == productId);
+
+                    if (existing.Quantity - quantity == 0)
+                    {
+                        current.ProductItems.Remove(existing);
+                    }
+                    else
+                    {
+                        existing.Quantity -= quantity;
+                        existing.UpdatedAt = removedAt;
+                    }
+
+                    break;
+                case (Confirmed (var clientId, var confirmedAt), not null):
+                    current.Status = ShoppingCartModel.ShoppingCartStatus.Confirmed;
+                    current.ClientId = clientId;
+                    current.ConfirmedAt = confirmedAt;
+                    dbContext.ShoppingCarts.Update(current);
+                    break;
+                case (Cancelled, not null):
+                    dbContext.ShoppingCarts.Remove(current);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(@event));
+            }
+        }
+
+        return dbContext.SaveChangesAsync(ct);
+    }
+}
+
+public interface IProductPriceCalculator
+{
+    IReadOnlyList<PricedProductItem> Calculate(params ProductItem[] productItems);
+}
+
+public class DummyProductPriceCalculator(decimal price): IProductPriceCalculator
+{
+    public IReadOnlyList<PricedProductItem> Calculate(params ProductItem[] productItems) =>
+        productItems.Select(pi => new PricedProductItem(pi.ProductId, pi.Quantity, price)).ToArray();
+}
+
+public class Test
+{
+    [Fact]
+    public void GivenEmptyShoppingCart_WhenAddProduct_ThenReturnsOpenedAndProductAddedEvents()
+    {
+        //Given
+        var price = (decimal)new Random().NextDouble() * 100m;
+        var now = DateTimeOffset.Now;
+
+        var cardId = Guid.NewGuid();
+        var productItem = new ProductItem(Guid.NewGuid(), 10);
+
+        var state = new Empty();
+
+        var decider = new ShoppingClassDecider(new DummyProductPriceCalculator(price), new FakeTimeProvider(now));
```

Comment:
How can you resist using the .NET 8 TimeProvider stuff :P

Reply:
Yeah, it's redundant, I just wanted to show people how to use it. Or to explain why not to use it.